request_helpers_test.go

```diff
@@ -0,0 +1,249 @@
+	NestedString string `json:"nested_string"`
+}
+
+type testResultType struct {
+	ID   int                  `json:"id"`
+	Foo  string               `json:"foo"`
+	Cool testResultNestedType `json:"cool"`
+}
+
+var testResponse = testResultType{
+	Foo: "test",
+	ID:  123,
+	Cool: testResultNestedType{
+		NestedInt:    456,
+		NestedString: "test2",
+	},
+}
+
+func TestRequestHelpers_get(t *testing.T) {
+	client := testutil.CreateMockClient(t, NewClient)
+
+	httpmock.RegisterRegexpResponder("GET", testutil.MockRequestURL("/foo/bar"),
+		httpmock.NewJsonResponderOrPanic(200, &testResponse))
+
+	result, err := doGETRequest[testResultType](
+		context.Background(),
+		client,
+		"/foo/bar",
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !reflect.DeepEqual(*result, testResponse) {
+		t.Fatalf("actual response does not equal desired response: %s", cmp.Diff(result, testResponse))
+	}
+}
+
+func TestRequestHelpers_post(t *testing.T) {
+	client := testutil.CreateMockClient(t, NewClient)
+
+	httpmock.RegisterRegexpResponder("POST", testutil.MockRequestURL("/foo/bar"),
+		testutil.MockRequestBodyValidate(t, testResponse, testResponse))
+
+	result, err := doPOSTRequest[testResultType](
+		context.Background(),
+		client,
+		"/foo/bar",
+		testResponse,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !reflect.DeepEqual(*result, testResponse) {
+		t.Fatalf("actual response does not equal desired response: %s", cmp.Diff(result, testResponse))
+	}
+}
+
+func TestRequestHelpers_put(t *testing.T) {
+	client := testutil.CreateMockClient(t, NewClient)
+
+	httpmock.RegisterRegexpResponder("PUT", testutil.MockRequestURL("/foo/bar"),
+		testutil.MockRequestBodyValidate(t, testResponse, testResponse))
+
+	result, err := doPUTRequest[testResultType](
+		context.Background(),
+		client,
+		"/foo/bar",
+		testResponse,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !reflect.DeepEqual(*result, testResponse) {
+		t.Fatalf("actual response does not equal desired response: %s", cmp.Diff(result, testResponse))
+	}
+}
+
+func TestRequestHelpers_delete(t *testing.T) {
+	client := testutil.CreateMockClient(t, NewClient)
+
+	httpmock.RegisterRegexpResponder(
+		"DELETE",
+		testutil.MockRequestURL("/foo/bar/foo%20bar"),
+		httpmock.NewStringResponder(200, "{}"),
+	)
+
+	if err := doDELETERequest(
+		context.Background(),
+		client,
+		formatAPIPath("/foo/bar/%s", "foo bar"),
+	); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestRequestHelpers_paginateAll(t *testing.T) {
+	client := testutil.CreateMockClient(t, NewClient)
+
+	numRequests := 0
+
+	httpmock.RegisterRegexpResponder(
+		"GET",
+		testutil.MockRequestURL("/foo/bar"),
+		mockPaginatedResponse(
+			buildPaginatedEntries(12),
+			&numRequests,
+		),
+	)
+
+	response, err := getPaginatedResults[testResultType](
+		context.Background(),
+		client,
+		"/foo/bar",
+		&ListOptions{
+			PageSize: 4,
+			Filter:   "{\"foo\": \"bar\"}",
+		},
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if numRequests != 3 {
```

Comment:
Good question! Since the mock result has a total of 12 entries (line 126) and each page has 4 entries(line 136), we expect the helper make three requests (pages 1, 2, and 3)

Reply:
Ah got it. Thank you for explaining it!